ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java

```diff
@@ -98,27 +106,44 @@ public void execute(Input input, ActionListener<Output> listener) {
                         try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, r.getSourceAsBytesRef())) {
                             ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
                             MLAgent mlAgent = MLAgent.parse(parser);
-                            ActionListener<Object> agentActionListener = ActionListener.wrap(output -> {
-                                if (output != null) {
-                                    Gson gson = new Gson();
-                                    if (output instanceof ModelTensorOutput) {
-                                        ModelTensorOutput modelTensorOutput = (ModelTensorOutput) output;
-                                        modelTensorOutput.getMlModelOutputs().forEach(outs -> {
-                                            for (ModelTensor mlModelTensor : outs.getMlModelTensors()) {
-                                                modelTensors.add(mlModelTensor);
-                                            }
-                                        });
-                                    } else if (output instanceof ModelTensor) {
-                                        modelTensors.add((ModelTensor) output);
-                                    } else if (output instanceof List) {
-                                        if (((List) output).get(0) instanceof ModelTensor) {
-                                            ((List<ModelTensor>) output).forEach(mlModelTensor -> modelTensors.add(mlModelTensor));
-                                        } else if (((List) output).get(0) instanceof ModelTensors) {
-                                            ((List<ModelTensors>) output).forEach(outs -> {
+                            String memoryType = mlAgent.getMemory().getType();
+                            String memoryId = inputDataSet.getParameters().get(MEMORY_ID);
+                            String appType = mlAgent.getAppType();
+                            String title = inputDataSet.getParameters().get(QUESTION);
+
+                            ConversationIndexMemory.Factory conversationIndexMemoryFactory =
+                                (ConversationIndexMemory.Factory) memoryFactoryMap.get(memoryType);
```

Comment:
does this mean we can only use `conversation_index` as memory type for all agents?

Reply:
In my opinion, memory should not be tied to an agent and should be common to all agents. We could have different implementations for memory that could be dynamically chosen but it should reusable across all agents.

Reply:
@jngz-es @ylwu-amzn have question about `Memory` interface, is memory instance a single memory object or it's a memory container? 

[`BufferedMemory`](https://github.com/opensearch-project/ml-commons/blob/feature/agent_framework_dev/ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/BufferMemory.java) implements like a container for all sessions

[`ConversationIndexMemory`](https://github.com/opensearch-project/ml-commons/blob/feature/agent_framework_dev/ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java) have a property `conversation_id` associate with it, it's more like a single memory object

Reply:
Please ignore BufferedMemory for now, we don't use it right now as it is not P0 I think.
Yeah, ConversationIndexMemory is an object which is allocated for each request.